File: Algorithms_Part_2/Ex43HuffmanCompression.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.BST;
import edu.princeton.cs.algs4.BinaryIn;
import edu.princeton.cs.algs4.BinaryOut;
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.MinPQ;
import edu.princeton.cs.algs4.StdOut;

public class Ex43HuffmanCompression {

    private static final int R = 256;

    private static class Node implements Comparable<Node> {
        private final char key;
        private final Node left;
        private final Node right;
        private final int freq;

        public Node(char key, Node left, Node right, int freq) {
            this.key = key;
            this.left = left;
            this.right = right;
            this.freq = freq;
        }

        public boolean isLeafNode() {
            return left == null && right == null;
        }

        public int compareTo(Node that) {
            return Integer.compare(freq, that.freq);
        }
    }

    public void compress(String input, String output) {

        BinaryIn in = new BinaryIn(input);
        BinaryOut out = new BinaryOut(output);

        int chars = 0;
        int[] freq = new int[R];
        while (!in.isEmpty()) {
            chars++;
            freq[in.readChar()]++;
        }

        MinPQ<Node> st = new MinPQ<>();
        for (int i = 0; i < R; i++) {
            if (freq[i] > 0)
                st.insert(new Node((char) i, null, null, freq[i]));
        }

        while (st.size() > 1) {
            Node left = st.delMin();
            Node right = st.delMin();
            st.insert(new Node('.', left, right, left.freq + right.freq));
        }
        Node root = st.delMin();

        BST<Character, String> table = writeTrie(out, root);
        out.write(chars);
        /* StdOut.println("table:");
        for (Character ch : table.keys())
            StdOut.println((ch == '\n' ? "newLine" : ch) + ": " + table.get(ch));*/
        in = new BinaryIn(input);
        while (!in.isEmpty()) {

            char c = in.readChar();
            String s = table.get(c);
            // StdOut.print("char:" + (c == '\n' ? "newLine" : c) + " table:" + s + " = ");
            for (char ch : s.toCharArray()) {
                boolean bit = ch == '1';
                out.write(bit);
                // StdOut.print(bit);
            }
            //    StdOut.println();
        }

        out.close();

    }

    public void expand(String input, String output) {

        BinaryIn in = new BinaryIn(input);
        BinaryOut out = new BinaryOut(output);

        Node root = readTrie(in);
        int chars = in.readInt();
        Node node = root;
        while (!in.isEmpty()) {
            boolean bit = in.readBoolean();
            if (bit) node = node.right;
            else node = node.left;
            if (node.isLeafNode()) {
                out.write(node.key);
                node = root;
                chars--;
                if (chars == 0) break;
            }
        }

        out.close();
    }

    private Node readTrie(BinaryIn in) {
        Node node;
        boolean bit = in.readBoolean();
        if (bit) {
            char ch = in.readChar(8);
            node = new Node(ch, null, null, 0);
            // StdOut.println("char read:" + ch);
        }
        else node = new Node('.', readTrie(in), readTrie(in), 0);
        return node;
    }

    private BST<Character, String> writeTrie(BinaryOut out, Node root) {
        BST<Character, String> table = new BST<>();
        writeTrie(root, out, table, "");
        return table;
    }

    private void writeTrie(Node node, BinaryOut out, BST<Character, String> table, String prefix) {
        if (node == null) return;

        if (node.isLeafNode()) {
            out.write(true);
            out.write(node.key, 8);
            table.put(node.key, prefix);
            char c = node.key;
            //   StdOut.println("key:" + (c == '\n' ? "newLine" : c) + " prefix:" + prefix);
        }
        else {
            out.write(false);
            writeTrie(node.left, out, table, prefix + '0');
            writeTrie(node.right, out, table, prefix + '1');
        }
    }

    private static void binaryDump(int bitsPerLine, String filename) {

        BinaryIn in = new BinaryIn(filename);

        int count, line = 0;
        StdOut.print(line + ": ");
        for (count = 0; !in.isEmpty(); count++) {
            if (bitsPerLine == 0) {
                in.readBoolean();
                continue;
            }
            else if (count != 0 && count % bitsPerLine == 0) {
                StdOut.println();
                StdOut.print(++line + ": ");
            }
            if (in.readBoolean()) StdOut.print(1);
            else StdOut.print(0);
        }
        if (bitsPerLine != 0) StdOut.println();
        StdOut.println(count + " bits");
    }

    public static void main(String[] args) {
        Ex43HuffmanCompression compression = new Ex43HuffmanCompression();
        compression.compress("inputFile.txt", "outputFile.txt");
        compression.expand("outputFile.txt", "inputFileExpanded.txt");

        binaryDump(200, "inputFile.txt");
        binaryDump(200, "inputFileExpanded.txt");
        binaryDump(72, "outputFile.txt");

        try {
            StdOut.print("Original: " + (new In("inputFile.txt").readAll()));
        }
        catch (IllegalArgumentException e) {
            StdOut.println("input file is not present");
        }
        try {
            StdOut.println("Output  : " + (new In("outputFile.txt").readAll()));
        }
        catch (IllegalArgumentException e) {
            StdOut.println("output file is not present");
        }
        try {
            StdOut.print("Expanded: " + (new In("inputFileExpanded.txt").readAll()));
        }
        catch (IllegalArgumentException e) {
            StdOut.println("expanded file is not present");
        }

        try {
            StdOut.println("isSame?" + (new In("inputFile.txt").readAll())
                    .equals(new In("inputFileExpanded.txt").readAll()));
        }
        catch (IllegalArgumentException e) {

        }

    }
}

File: PythonLearnig/Algorithms_Part_2/Ex43HuffmanCompression.py
from Ex45BinaryIn import BinaryIn
from Ex46BinaryOut import BinaryOut
from Ex15KruskalsMST import MinPQ

R = 256


class Node:

    def __init__(self, ch, freq, left, right):
        self.ch = ch
        self.freq = freq
        self.left = left
        self.right = right

    def is_leaf_node(self):
        return self.left == self.right is None

    def compare_to(self, that):
        return self.freq - that.freq


def compress(input_name, output_name):

    freq = [0] * R
    with open(input_name) as f:
        for c in f.read():
            freq[ord(c)] += 1
    N = sum(freq)

    root = __compute_trie(freq)

    i_stream = BinaryIn(input_name)
    o_stream = BinaryOut(output_name)

    st = {}
    __write_trie__(o_stream, root, st, [])
    o_stream.write_int_32(N)

    while not i_stream.is_empty():
        ch = i_stream.read_char_r(8)
        for bit in st[ch]:
            o_stream.write_boolean(bit)

    i_stream.close()
    o_stream.close()


def expand(input_name, output_name):

    i_stream = BinaryIn(input_name)
    o_stream = BinaryOut(output_name)

    root = __read_trie__(i_stream)
    N = i_stream.read_int()

    for _ in range(N):

        node = root
        while not node.is_leaf_node():

            bit = i_stream.read_boolean()
            if bit:
                node = node.right
            else:
                node = node.left

        o_stream.write_char_r(node.ch, 8)

    i_stream.close()
    o_stream.close()


def __compute_trie(freq):

    min_pq = MinPQ()
    for i in range(R):
        if freq[i] > 0:
            min_pq.put(Node(chr(i), freq[i], None, None))

    while min_pq.size() > 1:
        left = min_pq.delete_min()
        right = min_pq.delete_min()
        min_pq.put(Node('\0', left.freq + right.freq, left, right))

    return min_pq.delete_min()


def __write_trie__(o_stream, node, st, key):
    if node is None:
        return

    if node.is_leaf_node():
        o_stream.write_boolean(True)
        o_stream.write_char_r(node.ch, 8)
        st[node.ch] = key
    else:
        o_stream.write_boolean(False)
        __write_trie__(o_stream, node.left, st, key+[0])
        __write_trie__(o_stream, node.right, st, key+[1])


def __read_trie__(i_stream):

    if i_stream.is_empty():
        return None

    bit = i_stream.read_boolean()

    if bit:
        ch = i_stream.read_char_r(8)
        node = Node(ch, 0, None, None)
    else:
        left = __read_trie__(i_stream)
        right = __read_trie__(i_stream)
        node = Node('\0', 0, left, right)

    return node


def test_huffman_compression():

    input_file_name = 'input_file.txt'
    compressed_file_name = 'compressed_file.txt'
    output_file_name = 'output_file.txt'

    compress(input_file_name, compressed_file_name)
    expand(compressed_file_name, output_file_name)

    def read_text(file_name):
        try:
            with open(file_name) as f:
                return f.read()
        except FileNotFoundError:
            return ''

    original = read_text(input_file_name)
    compressed = read_text(compressed_file_name)
    output = read_text(output_file_name)

    print()
    print(f'Original   : {original}')
    print(f'Compressed : {compressed}')
    print(f'Output     : {output}')
    print()
    print(f'Is same? {original == output}')


if __name__ == '__main__':
    test_huffman_compression()
